boskos/boskos.go

```diff
@@ -0,0 +1,306 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+)
+
+var configPath = flag.String("config", "projects.json", "Path to init projects file")
+var storage = flag.String("storage", "", "Path to presistent volume to load the state")
+
+type GCPProject struct {
+	Name     string        `json:"name"`
+	Owner    string        `json:"owner"`
+	Duration time.Duration `json:"duration"`
+	Start    time.Time     `json:"start"`
+}
+
+func main() {
+	flag.Parse()
+	logrus.SetFormatter(&logrus.JSONFormatter{})
+
+	r := NewRancho(*configPath, *storage)
+	if r == nil {
+		logrus.Fatal("Fail to create my ranch!")
+	}
+
+	for _, proj := range r.MyCattle {
+		logrus.Infof("Donated Project : %v", proj.Name)
+	}
+
+	http.Handle("/", handleDefault())
+	http.Handle("/request", handleRequestProject(r))
+	http.Handle("/return", handleReturnProject(r))
+	http.Handle("/sync", handleHeartbeat(r))
+
+	logrus.Info("Start Service")
+	logrus.WithError(http.ListenAndServe(":8080", nil)).Fatal("ListenAndServe returned.")
+}
+
+func handleDefault() http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleDefault").Infof("From %v", req.RemoteAddr)
+	}
+}
+
+func handleRequestProject(r *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleRequestProject").Infof("From %v", req.RemoteAddr)
+
+		// validate vars
+		job := req.URL.Query().Get("job")
+		durationStr := req.URL.Query().Get("duration")
+		if job == "" || durationStr == "" {
+			logrus.Warning("[BadRequest]Job and duration must present in the request")
+			http.Error(res, "Job and duration must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		duration, err := time.ParseDuration(durationStr)
+		if err != nil {
+			http.Error(res, "Invalid Duration", http.StatusBadRequest)
+			logrus.WithError(err).Infof("Bad duration: %v", durationStr)
+			return
+		}
+
+		logrus.Infof("Job %v request for a project for %v", job, durationStr)
+
+		proj, err := r.GimmeCattle(job, duration)
+		if err != nil {
+			logrus.WithError(err).Info("Error from GimmeCattle")
+			http.Error(res, fmt.Sprintf("Failed request with error %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		if proj != nil {
+			projJSON, err := json.Marshal(proj)
+			if err != nil {
+				logrus.WithError(err).Infof("json.Marshal failed : %v", proj)
+				http.Error(res, error.Error(err), http.StatusInternalServerError)
```

Comment:
`error.Error(err)` what is that supposed to do? It should probably be `err.Error()`.